### ISSUE TITLE
How to install this?

### ISSUE THREAD

**Issue by philipsd6** [ministryofjustice/salt-cloud-provider-vcloud#5]
It's not clear how this package should be installed to work with an existing Salt master. After installing it via `python ./setup.py install` I can load it in an interactive python session:

``` pycon
Python 2.7.3 (default, Feb 27 2014, 19:58:35)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from salt.cloud.clouds import vcloud
>>>
```

However, `salt-cloud` does not find it. If I manually run `cp -pr salt-cloud-provider-vcloud/salt/cloud/clouds extmods/` to copy this into my extmods directory, vcloud can't load vcloud_support:

``` traceback
[DEBUG   ] Failed to import cloud vcloud, this is most likely NOT a problem:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/loader.py", line 875, in load_modules
    ), fn_, path, desc
  File "/etc/salt/extmods/clouds/vcloud.py", line 15, in <module>
    from .vcloud_support.libcloud_vcloud import get_vcloud_connection, wait_for_private_ips,\
ImportError: No module named vcloud_support.libcloud_vcloud
```

**Comment by ashb** [comment]
I remember this was a bit ... rough because of how salt loads providers.

Which version of salt cloud are you testing against? I think if you are on 2014.1 then you can just do `pip install git+http://github.com/ministryofjustice/salt-cloud-provider-vcloud.git#egg=salt-cloud-provider-vcloud`

I think.

(Salt changed paths between 2014 and 0.17. Also it looks at files in _one_ dir rather than in the load path globally so is quite picky about where it gets installed.)

**Comment by ashb** [comment]
In the interactive session when running setup.py what does this give you:

``` python
import salt.cloud.clouds.vcloud
salt.cloud.clouds.vcloud.__file__
```

Alternatively what is is /etc/salt/extmods/clouds when you install it that way.

**Comment by philipsd6** [comment]
I'm testing with git develop, so this is the (b)leading edge. Salt is installed in `/usr/lib/python2.7/dist-packages/salt`

I ran `pip install git+http://github.com/ministryofjustice/salt-cloud-provider-vcloud.git#egg=salt-cloud-provider-vcloud`

This installs to:

``` console
# ls -1 /usr/local/lib/python2.7/dist-packages/
apache_libcloud-0.14.1.egg-info
libcloud
salt
salt_cloud_provider_vcloud-0.1.1_alpha.egg-info
salt_cloud_provider_vcloud-0.1.1_alpha-nspkg.pth
```

But Salt is still not finding it:

```
# salt-cloud --list-providers
[INFO    ] salt-cloud starting
[WARNING ] The cloud driver, 'vcloud', configured under the 'devcloud' cloud provider alias was not loaded since 'vcloud.get_configured_provider()' could not be found. Removing it from the available providers list
Error: There was an error listing providers: There are no cloud providers configured.
```

Interactive session:

``` pycon
Python 2.7.3 (default, Feb 27 2014, 19:58:35)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import salt.cloud.clouds.vcloud
>>> salt.cloud.clouds.vcloud.__file__
'/usr/local/lib/python2.7/dist-packages/salt_cloud_provider_vcloud-0.1.1_alpha-py2.7.egg/salt/cloud/clouds/vcloud.pyc'
```

And now I seem to have entered namespace hell, since running pip uninstall and re-installing it a couple times:

``` console
# salt-cloud -l all --list-providers
Traceback (most recent call last):
  File "/usr/bin/salt-cloud", line 6, in <module>
    from salt.scripts import salt_cloud
ImportError: No module named scripts
# pip uninstall -y salt-cloud-provider-vcloud
Traceback (most recent call last):
  File "/usr/bin/pip", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2727, in <module>
    add_activation_listener(lambda dist: dist.activate())
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 700, in subscribe
    callback(dist)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2727, in <lambda>
    add_activation_listener(lambda dist: dist.activate())
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 2230, in activate
    map(declare_namespace, self._get_metadata('namespace_packages.txt'))
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1811, in declare_namespace
    declare_namespace(parent)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1811, in declare_namespace
    declare_namespace(parent)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1827, in declare_namespace
    _handle_ns(packageName, path_item)
  File "/usr/lib/python2.7/dist-packages/pkg_resources.py", line 1797, in _handle_ns
    loader.load_module(packageName); module.__path__ = path
  File "/usr/lib/python2.7/pkgutil.py", line 246, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "/usr/lib/python2.7/dist-packages/salt/__init__.py", line 37, in <module>
    from salt.utils import migrations
  File "/usr/lib/python2.7/dist-packages/salt/utils/__init__.py", line 94, in <module>
    import salt.minion
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 60, in <module>
    import salt.client
  File "/usr/lib/python2.7/dist-packages/salt/client/__init__.py", line 31, in <module>
    import salt.config
  File "/usr/lib/python2.7/dist-packages/salt/config.py", line 38, in <module>
    import salt.cloud.exceptions
ImportError: No module named exceptions
```

**Comment by philipsd6** [comment]
Looks like somehow `/usr/local/lib/` is being loaded before `/usr/lib` now:

``` console
# python -m site
sys.path = [
    '/root',
    '/usr/lib/python2.7',
    '/usr/lib/python2.7/plat-linux2',
    '/usr/lib/python2.7/lib-tk',
    '/usr/lib/python2.7/lib-old',
    '/usr/lib/python2.7/lib-dynload',
    '/usr/local/lib/python2.7/dist-packages',
    '/usr/lib/python2.7/dist-packages',
    '/usr/lib/python2.7/dist-packages/PIL',
    '/usr/lib/python2.7/dist-packages/gst-0.10',
    '/usr/lib/python2.7/dist-packages/gtk-2.0',
    '/usr/lib/pymodules/python2.7',
```

**Comment by shellhacks** [comment]
Hello,
Were you able to fix the issue?

As i met the same error and it seems that salt-cloud just doesn't load the vcloud module at all:

```
# salt-cloud -l all --list-providers[DEBUG   ] Reading configuration from /etc/salt/cloud
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Missing configuration file: /etc/salt/cloud.providers
[DEBUG   ] Including configuration from '/etc/salt/cloud.providers.d/vcloud.conf'
[DEBUG   ] Reading configuration from /etc/salt/cloud.providers.d/vcloud.conf
[DEBUG   ] Missing configuration file: /etc/salt/cloud.profiles
[DEBUG   ] Including configuration from '/etc/salt/cloud.profiles.d/vcloud.conf'
[DEBUG   ] Reading configuration from /etc/salt/cloud.profiles.d/vcloud.conf
[TRACE   ] None of the required configuration sections, 'logstash_udp_handler' and 'logstash_zmq_handler', were found the in the configuration. Not loading the Logstash logging handlers module.
[DEBUG   ] Configuration file path: /etc/salt/cloud
[TRACE   ] Trying pysss.getgrouplist for 'root'
[TRACE   ] Trying generic group list for 'root'
[TRACE   ] Group list for user 'root': []
[INFO    ] salt-cloud starting
[TRACE   ] 'parallels.avail_sizes' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] Failed to import clouds vcloud:
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/salt/loader.py", line 951, in _load_module
    ), fn_, fpath, desc)
  File "/usr/lib/python2.6/site-packages/salt/cloud/clouds/vcloud.py", line 15, in <module>
    from .vcloud_support.libcloud_vcloud import get_vcloud_connection, wait_for_private_ips,\
ImportError: No module named vcloud_support.libcloud_vcloud
[DEBUG   ] Failed to import clouds joyent:
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/salt/loader.py", line 951, in _load_module
    ), fn_, fpath, desc)
  File "/usr/lib/python2.6/site-packages/salt/cloud/clouds/joyent.py", line 68, in <module>
    from Crypto.Signature import PKCS1_v1_5
ImportError: No module named Signature
[DEBUG   ] Could not LazyLoad parallels.avail_sizes
[TRACE   ] 'parallels.avail_locations' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] LazyLoaded parallels.avail_locations
[TRACE   ] 'proxmox.avail_sizes' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] LazyLoaded proxmox.avail_sizes
[TRACE   ] 'saltify.destroy' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] Could not LazyLoad saltify.destroy
[TRACE   ] 'saltify.avail_sizes' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] Could not LazyLoad saltify.avail_sizes
[TRACE   ] 'saltify.avail_images' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] Could not LazyLoad saltify.avail_images
[TRACE   ] 'saltify.avail_locations' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] Could not LazyLoad saltify.avail_locations
[TRACE   ] 'rackspace.reboot' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] LazyLoaded rackspace.reboot
[TRACE   ] 'openstack.list_locations' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] LazyLoaded openstack.list_locations
[TRACE   ] 'rackspace.list_locations' has been marked as not supported. Removing from the list of supported cloud functions
[DEBUG   ] LazyLoaded rackspace.list_locations
[DEBUG   ] Could not LazyLoad vcloud.get_configured_provider
[WARNING ] The cloud driver, 'vcloud', configured under the 'my_vcloud' cloud provider alias was not loaded since 'vcloud.get_configured_provider()' could not be found. Removing it from the available providers list
Error: There was an error listing providers: There are no cloud providers configured.
```

Salt version:

```
# salt-cloud --versions-report
            Salt: 2015.5.0
          Python: 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
          Jinja2: 2.2.1
        M2Crypto: 0.20.2
  msgpack-python: 0.4.4
    msgpack-pure: Not Installed
        pycrypto: 2.0.1
         libnacl: Not Installed
          PyYAML: 3.10
           ioflo: Not Installed
           PyZMQ: 14.3.1
            RAET: Not Installed
             ZMQ: 3.2.5
            Mako: Not Installed
 Apache Libcloud: 0.17.0
```

Installation was performed with:

```
# pip install git+http://github.com/ministryofjustice/salt-cloud-provider-vcloud.git#egg=salt-cloud-provider-vcloud
```

Provider settings:

```
# cat /etc/salt/cloud.profiles.d/vcloud.conf
vcd_tiny_vdc_a:
  provider: my_vcloud
  image: vcloud-tpl-centos6-64b
  size: tiny
  vdc: 'my-org-a (BASIC)'

# cat /etc/salt/cloud.providers.d/vcloud.conf 
my_vcloud:
  provider: vcloud
  user: admin
  org: org
  secret: pass
  host: vcloud.examplr.tld

  private_key: /root/minion-ssh.pem
  ssh_username: provisioning
  display_ssh_output: True

# cat /etc/salt/cloud.maps.d/vcloud.conf 
vcdtiny_vdc_a:
  - loadbalancer.myapp:
      network: Internet
      # These are normal cloud.map settings and nothing to do with this
      # provider.
      minion:
        log_level: debug
      grains:
        roles:
          - loadbalancer
```

Can load module manually:

```
# python
Python 2.6.6 (r266:84292, Jan 22 2014, 09:42:36) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import salt.cloud.clouds.vcloud
>>> salt.cloud.clouds.vcloud.__file__
'/usr/lib/python2.6/site-packages/salt/cloud/clouds/vcloud.pyc'
>>> 
```

Thanks in advance for help!

**Comment by philipsd6** [comment]
@shellhacks I haven't worked with this anymore. Salt's develop branch has newly added support for vmware, based on pyVmomi, which I've been using successfully to create VMs in our vCenter sandbox.

See [Getting Started With VMware](http://docs.saltstack.com/en/develop/topics/cloud/vmware.html#getting-started-with-vmware)

**Comment by shellhacks** [comment]
@philipsd6 Thanks for information. Unfortunately i need to configure salt-cloud exactly for vCloud Director. Hope @ashb keeps an eye on this issue and will leave a feedback either.

**Comment by ashb** [comment]
Hi @shellhacks,

Much like @philipsd6 we've also stopped using this (mainly cos we're not using VMware anymore) but I'll try.

Okay so taking a look and I think I remember the problem... but maybe not the solution.

Because of how salt loads the cloud files (it doesn't do a normal python import but instead looks at files on disk) it has to be installed into the exact same path as the existing salt-cloud providers.

For me on an Ubuntu 14.04 VM the salt .deb packages installed things in /usr/lib/python2.7/dist-packages/salt/cloud/clouds  and pip put things under /usr/local/lib/.. instead.

This seemed to do the trick for me:

```
pip install --install-option="--install-purelib=/usr/lib/python2.7/dist-packages" \
    git+http://github.com/ministryofjustice/salt-cloud-provider-vcloud.git
```

but I don't think you had that problem judging from the errors. `/usr/lib/python2.6/site-packages/salt/cloud/clouds/` should have about 50 files in.

I think I have a fix though! Something about how salt loads the python module changed along the way so the relative import I was doing doesn't work anymore.

If you do `pip uninstall salt-cloud-provider-vcloud` and then reinstall with the try-to-fix-install branch for me this was done via this command:

``` bash
pip install --install-option="--install-purelib=/usr/lib/python2.7/dist-packages" \
    git+http://github.com/ministryofjustice/salt-cloud-provider-vcloud.git@try-to-fix-install
```

Note the "@try-to-fix-install" part on the end. You can probably omit the `--install-option` part but I needed it.

After doing that I got _some_ output from your sample config which is better than an error!

**Comment by shellhacks** [comment]
Hello @ashb, thanks for feedback. I was able to fix the issue with the relative path. I am new on github.com and don't know yet how to create commits here, but here is the diff that worked for me:

```
--- /usr/lib/python2.6/site-packages/salt/cloud/clouds/vcloud.py.orig    2015-05-28 14:56:17.272122580 +0200
+++ /usr/lib/python2.6/site-packages/salt/cloud/clouds/vcloud.py    2015-05-28 14:52:40.557126494 +0200
@@ -12,7 +12,7 @@
 import libcloud
 from libcloud.compute.base import NodeImage

-from .vcloud_support.libcloud_vcloud import get_vcloud_connection, wait_for_private_ips,\
+from vcloud_support.libcloud_vcloud import get_vcloud_connection, wait_for_private_ips,\
     lookup_gateway_info, apply_nat_rules, parse_nat_rules, create_vm

 # Get logging started
```

Here is the structure of installed files:

```
# pip show -f salt-cloud-provider-vcloud
---
Name: salt-cloud-provider-vcloud
Version: 0.1.1-alpha
Location: /usr/lib/python2.6/site-packages
Requires: pyyaml, jinja2, lxml, apache-libcloud, salt
Files:
  ../salt/cloud/clouds/vcloud.py
  ../salt/cloud/clouds/vcloud_support/__init__.py
  ../salt/cloud/clouds/vcloud_support/libcloud_vcloud.py
  ../salt/cloud/clouds/vcloud_support/nat_rules.py
  ../salt/cloud/clouds/vcloud.pyc
  ../salt/cloud/clouds/vcloud_support/__init__.pyc
  ../salt/cloud/clouds/vcloud_support/libcloud_vcloud.pyc
  ../salt/cloud/clouds/vcloud_support/nat_rules.pyc
  ./
  requires.txt
  dependency_links.txt
  not-zip-safe
  PKG-INFO
  namespace_packages.txt
  SOURCES.txt
  top_level.txt
  ../salt_cloud_provider_vcloud-0.1.1_alpha-py2.6-nspkg.pth
```

Now salt-cloud at least loads the vcloud module:

```
# salt-cloud --list-providers
[INFO    ] salt-cloud starting
my-vcloud-config:
    ----------
    vcloud:
        ----------
```

And even shows the list of VMs with the below command:

```
# salt-cloud -f list_nodes vcloud
```

So i think that the issue with the relative path can be marked as fixed now.

But unfortunately now the process stops on the new stage... While trying to create a new VM, i receive the following error:

```
ValueError: Invalid attribute name u'xml:lang'
```

As it is a new error, i am going to create a new issue for it.
